### ISSUE TITLE
Massive package drops in mqtt connectors

### ISSUE THREAD

**Issue by w4tsn** [eclipse/ditto#858]
I can observe lots of package drops in the mqtt connectors (both mqtt and mqtt-5) lately. Accompanied by warnings in the logs only 10-20% of mqtt packages now make it through to ditto and get persistet. From the logs I suppose that the packages arrive at the connectivity stack but do not get persisted.

This problem first occurred in ditto 1.2. I'm currently operating ditto 1.3.x.

From the connectivity service logs I observe a lot of HiveMQTT client warnings and I'm pretty sure that those did not appear before 1.2:

```
connectivity_1   | 2020-10-26 16:54:58,112 WARN  [] c.h.c.i.m.h.p.i.MqttIncomingPublishService  - No publish flow registered for MqttStatefulPublish{stateless=MqttPublish{topic=ditto/in, payload=226byte, qos=AT_MOST_ONCE, retain=false}, packetIdentifier=-1, dup=false, topicAlias=0, subscriptionIdentifiers=[]}.
connectivity_1   | 2020-10-26 16:54:58,112 WARN  [] c.h.c.i.m.h.p.i.MqttIncomingPublishService  - No publish flow registered for MqttStatefulPublish{stateless=MqttPublish{topic=ditto/in, payload=505byte, qos=AT_MOST_ONCE, retain=false}, packetIdentifier=-1, dup=false, topicAlias=0, subscriptionIdentifiers=[]}.
connectivity_1   | 2020-10-26 16:54:58,389 INFO  [e81d6548-c0b2-45c0-bd4e-92155a174506] o.e.d.s.m.c.a.ConciergeForwarderActor akka://ditto-cluster/user/connectivityRoot/conciergeForwarder - Forwarding signal with ID <8A5EF0FC400F19AA> and type <things.commands:modifyFeatureProperty> to concierge enforcer
connectivity_1   | 2020-10-26 16:54:58,389 INFO  [e81d6548-c0b2-45c0-bd4e-92155a174506] o.e.d.s.m.a.AcknowledgementAggregatorActor akka://ditto-cluster/system/sharding/connection/7/mqtt-5/pa/$a/c1/messageMappingProcessor/ackr3aa-e81d6548-c0b2-45c0-bd4e-92155a174506 - Starting to wait for all requested acknowledgements <[twin-persisted]> for a maximum duration of <PT1M5S>.
```

I've an extensive array of stored history of metrics from past months and can confirm via deployment and rollback times of ditto 1.1, 1.2 and 1.3 that this issue was introduced somewhere between 1.1 and 1.2. 

One more side-note: the metrics from the devops endpoint seem not to include those failed attempts as failures in a connection sense

EDIT:

I found those additional log entries but I don't know if they are related:

```
connectivity_1   | 2020-10-26 16:57:33,022 INFO  [] o.e.d.s.c.m.MessageMappingProcessorActor akka://ditto-cluster/system/sharding/connection/7/mqtt-5/pa/$a/c1/messageMappingProcessor - Got DittoRuntimeException 'json.field.missing' when ExternalMessage was processed: JSON did not include required </path> field! - Check if all required JSON fields were set.
connectivity_1   | 2020-10-26 16:58:09,766 INFO  [] o.e.d.s.c.m.MessageMappingProcessorActor akka://ditto-cluster/system/sharding/connection/7/mqtt-5-othermo-de/pa/$a/c1/messageMappingProcessor - Got DittoRuntimeException 'json.key.invalid' when ExternalMessage was processed: The JSON key <Vorlauf-Solltemperatur Anlagen / Kessel> is invalid! - Neither slashes nor any control characters are allowed.
```

**Comment by thjaeckle** [comment]
Based on this issue https://github.com/hivemq/hivemq-mqtt-client/issues/367 the warning `"No publish flow registered for MqttStatefulPublish"` means:
> The message means that the client received a message, but you did not register a callback and therefore the message is dropped.

Thanks for reporting, we'll look into it.
- HiveMQ client was updated in Ditto 1.3.0 to HiveMQ client 1.2.0
- Ditto 1.2.0 already contained a refactoring of having an abstraction for MQTT3 and MQTT5 common code, the `AbstractMqttConsumerActor` .. probably something broke with this refactoring

**Comment by w4tsn** [comment]
Sounds good, Thanks!

Meanwhile I've built a tiny node app translating the messages from the mqtt broker to websocket message and vice versa :sweat_smile:

**Comment by thjaeckle** [comment]
@w4tsn can you provide a repoducer?
We have integration tests which test MQTT connections and they don't face any issues.

**Comment by w4tsn** [comment]
Hmm. Not really a reproducer in that sense, but here's what I'm doing.

This is my mqtt-connector config. Doesn't really matter if it's `mqtt` or `mqtt-5`. It's a SSL connection though, so you'll need certificates for a test.

```json
{
    "targetActorSelection": "/system/sharding/connection",
    "headers": {
        "aggregate": false
    },
    "piggybackCommand": {
        "type": "connectivity.commands:createConnection",
        "connection": {
            "id": "mqtt-5-mosquitto-org",
            "name": "dropping MQTT Gateway",
            "connectionType": "mqtt-5",
            "connectionStatus": "open",
            "failoverEnabled": true,
            "uri": "ssl://test.mosquitto.org:8883",
            "validateCertificates": true,
            "ca": "REDACTED",
            "credentials": {
                "type": "client-cert",
                "cert": "REDACTED",
                "key": "REDACTED"
            },
            "sources": [
                {
                    "addresses": [
                        "ditto/in/#"
                    ],
                    "authorizationContext": ["nginx:ditto"],
                    "qos": 0,
                    "filters": []
                }
            ],
            "targets": [
                {
                    "address": "ditto/out/{{ topic:full }}",
                    "topics": [
                        "_/_/things/twin/events",
                        "_/_/things/live/messages"
                    ],
                    "authorizationContext": [
                        "nginx:ditto"
                    ],
                    "qos": 0
                }
            ]
        }
    }
}
```

The broker in use is a mosquitto broker with certificate based authentication and an ACL file managing permissions on the ditto/in and ditto/out topics.

**Comment by w4tsn** [comment]
@thjaeckle I still have no reproducer :S Do your integration tests cover percentage package drops? I suspect some kind of load issue, since some data goes through but the majority is dropped.

I now also know that if I turn off the newly introduced configuration options `reconnectForRedelivery` and `separatePublisherClient` the warnings about the missing publish flow go away. This has no influence on the package drops however, so I think this wasn't related to begin with. As a side-note on this: I'm using a TLS certificate connection and the broker is configured to only allow singular authorization per certificate so the option `separatePublisherClient` can't actually work if I understand this correctly. One of both clients is either dropped when the other connects or connection is denied for one of them, don't know for sure.

I'm operating a little work-around translating the mqtt-messages to ditto via websockets and vice-versa. This way I don't see the package drops although the load on the server and cluster did not change.

**Comment by w4tsn** [comment]
Update: OK, well, it seems that the problem was due to a polluted connectivity service. There were multiple connections trying to access the broker with the same credentials which lead to constant reconnects and thus package drops. 

Restarting the connectivity service solved the issue. I thought I had already tried that, but I suppose I did when those multiple connections were still present and didn't thought about it after actually removing all other connections.

In return that means somehow older connections survived the piggyback removal and stayed open / active until I restarted the service (I did not recreate the container). I've confirmed in mongodb that there were three connections in total, 2 marked as deleted, one of them as active. But I could observe multiple clients trying to connect until I restartet the connectivity service.

**Comment by thjaeckle** [comment]
> In return that means somehow older connections survived the piggyback removal and stayed open / active until I restarted the service (I did not recreate the container). I've confirmed in mongodb that there were three connections in total, 2 marked as deleted, one of them as active. But I could observe multiple clients trying to connect until I restartet the connectivity service.

Hi @w4tsn 
That's both good to hear and bad to hear :-D
Good that you fixed it and that is not a fundamental problem in Ditto's MQTT connectivity.
Not easy to reproduce this however as this is probably an update issue of an older Ditto version to a newer one.

If you find out how to reproduce we would be happy to look into this - that's the first time we saw an issue like that.

Regarding the warning and settings of `reconnectForRedelivery` and `separatePublisherClient` we will look into it.
Maybe the default should be not `true` for the both settings, but only when configured specifically.

**Comment by BobClaerhout** [comment]
I would like to add my 2 cents in this as well.
We also encountered this issue (the connection being kept alive multiple times). I don't have any way to reproduce this easily but I think the chances of encountering this issue is higher when you are creating faulty connections. I usually need a few iterations when creating a new connection to configure it correctly. During these re-configurations a connection might get "lost" in the sense that this connection will live on its own.
This results in multiple connections targeting the same (external) host. This host does not allow multiple connections with the same credentials resulting in disconnections all the time. Restarting the ditto-connectivity service ensures only 1 connection is created again which solves the problem.

**Comment by thjaeckle** [comment]
I could not reproduce with the current master which will soon be released as Ditto 2.0.
I guess this was fixed in some feature enhancement between 1.5 and 2.0
